### PR TITLE
Reload source tiles when a filter or layout property is modified

### LINKED ISSUE #5701: Setting a layout property needs to trigger tile reparse
When a layout property is modified, all existing tiles for that source need to be reparsed, so that vertex arrays that are calculated based on that property are recalculated.

This will fix https://github.com/mapbox/mapbox-gl-native/issues/5610#issuecomment-233004625.

Additionally, once data-driven-styling lands, modifying a data-driven paint property also needs to trigger a reparse.

We should port the general strategy for batched updates from gl-js -- see https://github.com/mapbox/mapbox-gl-js/pull/2355 and the current [style.js](https://github.com/mapbox/mapbox-gl-js/blob/master/js/style/style.js).

cc @ivovandongen @mourner @kkaefer

### LINKED ISSUE #6017: allow for more efficient batch runtime style API changes
We want a transaction block model such as `CATransaction` which will defer calls to `-[MGLBaseStyleLayer update]` until the transaction is completed, allowing for many setters to be called which are all made effective with one `-update` call. 

A model for design here might be a singleton which counts to `1` when the transaction `begin`s, decrements to `0` when the transaction `commit`s, and throws an exception if it is ever tried to be nested with a secondary or later `begin` (though, that would probably be harmless—`commit` could just always reset to `0`). 

/cc @frederoni @1ec5

### LINKED ISSUE #6063: Android runtime styling changes not updated when changing existing layer's filter
**Platform:** Android
**Mapbox SDK version:** Nightly

### Steps to trigger behavior

 1.  Load up a style url
 2.  Set a timer to change one of the existing layers' filter periodically.
(our case is highlighting different areas)
 
### Expected behavior
Styles update according to a specified filter

### Actual behavior
The initial render actually looks about right, but consecutive changes to the filter aren't visible on screen. Zooming in or out a full level seems to cause the current correct style to be rendered (and cached) on that zoom level, but again once all zoom levels are populated updates seem pretty erratic. Also sometimes it seems that things are only updated to a tile border.

I've tried invalidate() on MapView, MapboxMap, and individual layers, as workaround, but doesn't help. 

Apologies for the lack of code to demonstrate the issue, I just wanted to give you a heads up as quickly as possible about this issue.

### LINKED ISSUE #6116: [android] Setting a layout property needs to trigger tile reparse
This tracks the work for https://github.com/mapbox/mapbox-gl-native/issues/5701 on Android. Mainly to flag the issue in the release milestone and carry out Android specific tests.

/cc: @mapbox/android

### LINKED ISSUE #6233: Changing visibility needs to trigger tile reparse
This is a special case of #5701 but I want to file it separately so I can refer to the issue number in a test-suite regression test.

1. Load a style with line layers L1 (initially `"visibility": "visible"`) and L2 (initially `"visibility": "none"`) referencing the same source.
2. Wait for L1 to render. (This layer exists only to force the source to be loaded prior to step 3.)
3. Set L2's visibility to "visible".

### Expected behavior

L2 renders.

### Actual behavior

L2 does not render.

### PR DESCRIPTION
**Pull request by jfirebaugh**
Fixes #5701
Fixes #6063 
Fixes #6116
Fixes #6017
Fixes #6233

When a layout property is modified, all existing tiles for that source need to be reparsed, so that vertex arrays that are calculated based on that property are recalculated. The same is true for modifications to layer filters.

This PR can be divided into three parts:

* Setting up `Source`, `GeometryTile`, and `WorkerTile` so that tiles can be reloaded (first 5 commits). This involves ensuring that the worker can hold onto the tile data for reuse, rather than passing ownership back to the main thread. Because the main thread _also_ needs tile data for query purposes, it has to be made copyable. (/cc @kkaefer for review of this portion.)
* Setting up `Style` to be able to cause a reload at the appropriate time (next 2 commits). This required modifying our change notification infrastructure so that the style can observe layer mutations directly, rather than being indirectly notified in a way that requires the cooperation of SDK bindings. (/cc @1ec5 @frederoni @ivovandongen for review of this portion.)
* Connecting the two sides (final commit). This is relatively straightforward once the prerequisites are in place.

Remaining work:

* [x] `Style` should auto-batch source reloads
* [x] `NodeMap::SetFilter` should use `Layer::accept`. Or should we have an intermediate base class for `FillLayer`, `LineLayer`, `CircleLayer`, and `SymbolLayer`, with pure virtual common methods:
 ```
    virtual const std::string& getSourceID() const = 0;
    virtual const std::string& getSourceLayer() const = 0;
    virtual void setSourceLayer(const std::string& sourceLayer) = 0;

    virtual void setFilter(const Filter&) = 0;
    virtual const Filter& getFilter() const = 0;
 ```
* [ ] Fix race condition: `GeometryTile::redoLayout` assumes that the worker has tile data from a previous call to `worker.parseGeometryTile(...)`, but it also cancels the pending work request, which might be the very request that provides the worker with the tile data.
* [ ] Reload when GeoJSON source data changes.

### CODE REVIEW

**Comment by 1ec5** [comment]
* [ ] Remove the MGLBaseStyleLayer(MGLBaseStyleLayer_Private) category and delete MGLBaseStyleLayer_Private.h

**Comment by jfirebaugh** [comment]
There's a race condition in this implementation: `GeometryTile::redoLayout` assumes that the worker has tile data from a previous call to `worker.parseGeometryTile(...)`, but it also cancels the pending work request, which might be the very request that provides the worker with the tile data.

**Comment by ivovandongen** [comment]
@jfirebaugh Looking good! 

The examples with paint property updates work as expected now. I did notice however, that when removing a layer, nothing happens until I interact with the map (I would add a gif, but you wouldn't see my interaction anyway). You can try this out in the runtime style test activity if you like.

About the crash, I also don't have access yet (integrating the result in bitrise is underway iirc), but I ran into only one crash intermittently which @cammace reported yesterday: https://github.com/mapbox/mapbox-gl-native/issues/6193

File: src/mbgl/tile/vector_tile.cpp [hunk 1]
```diff
@@ -59,6 +59,10 @@ class VectorTileData : public GeometryTileData {
 public:
     VectorTileData(std::shared_ptr<const std::string> data);
 
+    std::unique_ptr<GeometryTileData> clone() const override {
+        return std::make_unique<VectorTileData>(*this);
+    }
+
```
Comment (kkaefer) [inline src/mbgl/tile/vector_tile.cpp]: Why are we making a copy here? `*TileData` objects should be designed in a read-only, all-const manner (all of its methods are `const`), so we should be able to use shared pointers. Given that they don't hold any resources, except for a `shared_ptr<const std::string>`, they can be destructed in any thread.

File: src/mbgl/util/ptr.hpp [hunk 2]
```diff
@@ -1,26 +0,0 @@
...
-        assert(*this);
-        return ::std::shared_ptr<T>::operator->();
-    }
-    auto operator*() const -> decltype(this->::std::shared_ptr<T>::operator*()) {
-        assert(*this);
-        return ::std::shared_ptr<T>::operator*();
-    }
-};
-} // namespace util
-} // namespace mbgl
```
Comment (kkaefer) [inline src/mbgl/util/ptr.hpp]: 👍

**Comment by kkaefer** [comment]
@jfirebaugh Why are we making a deep copy of the *TileData object? We should be able to use shared pointers for them.

**Comment by jfirebaugh** [comment]
@kkaefer `GeometryTileData` subclasses are not thread-safe -- in particular `VectorTileData`. And in general I don't want to introduce more cross-thread shared state (#667).

**Comment by kkaefer** [comment]
Ugh, I got fooled by the `const`ness of the member functions, but `VectorTileData` has `mutable` data members. We should probably remove the constness guarantee if they are mutable.

However, `*TileData` objects are not representing state; rather they are representing data, which should be immutable, and threadsafe.

**Comment by jfirebaugh** [comment]
If we want `*TileData` to represent immutable data, I think we need to guarantee that the implementations are either:

* Fully eager, and return const references to parsed data
* Fully lazy, and return values (potentially wrapped in `std::unique_ptr` but only for polymorphism)

Lazy-but-caching implementations with interior mutability don't mix well with cross-thread sharing.

@springmeyer, is this compatible with the plans you have for [vector-tile](https://github.com/mapbox/vector-tile)?

**Comment by ivovandongen** [comment]
@jfirebaugh @tobrun retrieved the logs for me, it seems there is another issue in play in this [testcase](https://github.com/mapbox/mapbox-gl-native/blob/fff0ebaa1754aabd9584c994624b7b948bec95af/platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/style/LineLayerTest.java#L250-L250): 

```
08-30 10:00:50.299: I/TestRunner(23699): started: testLineTranslateAnchor(com.mapbox.mapboxsdk.style.LineLayerTest)
08-30 10:00:50.302: I/MonitoringInstrumentation(23699): Activities that are still in CREATED to STOPPED: 1
08-30 10:00:50.303: D/ActivityInstrumentationRule(23699): Launching activity com.mapbox.mapboxsdk.testapp.activity.style.RuntimeStyleTestActivity
08-30 10:00:50.308: D/MonitoringInstrumentation(23699): execStartActivity(context, ibinder, ibinder, activity, intent, int, bundle
08-30 10:00:50.309: I/ActivityManager(834): START u0 {act=android.intent.action.MAIN flg=0x10000000 cmp=com.mapbox.mapboxsdk.testapp/.activity.style.RuntimeStyleTestActivity} from uid 10100 on display 0
08-30 10:00:50.318: V/WindowManager(834): addAppToken: AppWindowToken{1cd5935c token=Token{307c3dcf ActivityRecord{6d00e2e u0 com.mapbox.mapboxsdk.testapp/.activity.style.RuntimeStyleTestActivity t75}}} to stack=1 task=75 at 0
08-30 10:00:50.322: V/WindowManager(834): Adding window Window{3ac17c7 u0 Starting com.mapbox.mapboxsdk.testapp} at 5 of 9 (after Window{16f8cad7 u0 com.mapbox.mapboxsdk.testapp/com.mapbox.mapboxsdk.testapp.activity.style.RuntimeStyleTestActivity})
08-30 10:00:50.366: A/libc(23699): Fatal signal 11 (SIGSEGV), code 1, fault addr 0x18 in tid 24456 (Worker)
08-30 10:00:50.460: D/LifecycleMonitor(23699): Lifecycle status change: com.mapbox.mapboxsdk.testapp.activity.style.RuntimeStyleTestActivity@3b304617 in: STOPPED
08-30 10:00:50.461: D/mbgl(23699): {pboxsdk.testapp}[JNI]: nativeTerminateContext
08-30 10:00:50.461: D/mbgl(23699): {pboxsdk.testapp}[Android]: NativeMapView::terminateContext
08-30 10:00:50.461: D/mbgl(23699): {pboxsdk.testapp}[JNI]: nativeTerminateDisplay
08-30 10:00:50.461: D/mbgl(23699): {pboxsdk.testapp}[Android]: NativeMapView::terminateDisplay
08-30 10:00:50.489: I/DEBUG(359): *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
08-30 10:00:50.489: I/DEBUG(359): Build fingerprint: 'google/shamu/shamu:5.1/LMY47D/1743759:user/release-keys'
08-30 10:00:50.489: I/DEBUG(359): Revision: '33696'
08-30 10:00:50.489: I/DEBUG(359): ABI: 'arm'
08-30 10:00:50.489: I/DEBUG(359): pid: 23699, tid: 24456, name: Worker  >>> com.mapbox.mapboxsdk.testapp <<<
08-30 10:00:50.489: I/DEBUG(359): signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x18
08-30 10:00:50.539: I/DEBUG(359):     r0 00000018  r1 94bffb08  r2 0030b390  r3 791990d8
08-30 10:00:50.539: I/DEBUG(359):     r4 00000018  r5 94bffb08  r6 94bffc00  r7 94bffaf0
08-30 10:00:50.539: I/DEBUG(359):     r8 00000000  r9 94bffd58  sl 00000000  fp 94c7bf00
08-30 10:00:50.539: I/DEBUG(359):     ip a1adbee4  sp 94bffad8  lr a1a0cb49  pc b6e17682  cpsr 600f0030
08-30 10:00:50.539: I/DEBUG(359): backtrace:
08-30 10:00:50.539: I/DEBUG(359):     #00 pc 00017682  /system/lib/libc.so (pthread_mutex_lock+7)
08-30 10:00:50.539: I/DEBUG(359):     #01 pc 00379b45  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so (std::__ndk1::mutex::lock()+4)
08-30 10:00:50.539: I/DEBUG(359):     #02 pc 00136ecc  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.539: I/DEBUG(359):     #03 pc 00136e1c  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.539: I/DEBUG(359):     #04 pc 001342cc  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.539: I/DEBUG(359):     #05 pc 00134158  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.540: I/DEBUG(359):     #06 pc 0012ea1c  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.540: I/DEBUG(359):     #07 pc 00133d1c  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.540: I/DEBUG(359):     #08 pc 00133b2c  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.540: I/DEBUG(359):     #09 pc 00137074  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.540: I/DEBUG(359):     #10 pc 00136f58  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.540: I/DEBUG(359):     #11 pc 0012f0b0  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.540: I/DEBUG(359):     #12 pc 0012eff8  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so
08-30 10:00:50.540: I/DEBUG(359):     #13 pc 00016baf  /system/lib/libc.so (__pthread_start(void*)+30)
08-30 10:00:50.540: I/DEBUG(359):     #14 pc 00014af3  /system/lib/libc.so (__start_thread+6)
```

I could reproduce this locally only by running the entire test suite, the individual test case succeeds.

**Comment by jfirebaugh** [comment]
@ivovandongen Unfortunately the stack trace is missing symbols -- is it a release build?

**Comment by jfirebaugh** [comment]
Oh, I suppose #6221 is the cause of the missing symbols.

**Comment by springmeyer** [comment]
>> Lazy-but-caching implementations with interior mutability don't mix well with cross-thread sharing.

> @springmeyer, is this compatible with the plans you have for vector-tile?

Yes. Overall looks compatible. Also, I removed the `mutable` members in `vector-tile` with the goal of truly being able to pass VectorTileObjects as const and immutable.

**Comment by jfirebaugh** [comment]
@ivovandongen I think the lack of repaint when removing a layer is a separate issue; can you file a ticket?

Also can you post the new Android test output? Hopefully it has symbols in the stack trace now.

**Comment by ivovandongen** [comment]
> I think the lack of repaint when removing a layer is a separate issue; can you file a ticket?

I will. But I mentioned it here because this I hadn't seen this before this branch.

> Also can you post the new Android test output? Hopefully it has symbols in the stack trace now.

Yes. However, all tests pass now for me locally (ran them twice to be sure). I "symbolized" against the current state of your branch, so it might be off a little.

```
********** Crash dump: **********
Build fingerprint: 'google/shamu/shamu:5.1/LMY47D/1743759:user/release-keys'
pid: 23699, tid: 24456, name: Worker  >>> com.mapbox.mapboxsdk.testapp <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x18
Stack frame #00 pc 00017682  /system/lib/libc.so (pthread_mutex_lock+7)
Stack frame #01 pc 00379b45  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so (std::__ndk1::mutex::lock()+4): Routine sentry at /Volumes/Android/buildbot/src/android/ndk-r12-release/ndk/sources/cxx-stl/llvm-libc++/libcxx/include/ostream:233
Stack frame #02 pc 00136ecc  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine bool std::__ndk1::__insertion_sort_incomplete<mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::$_1&, mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::ID*>(mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::ID*, mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::ID*, mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::$_1&) at /Users/ivo/git/mapbox-gl-native/mason_packages/osx-x86_64/android-ndk/arm-9-r12b/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/algorithm:3750
Stack frame #03 pc 00136e1c  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine operator() at /Users/ivo/git/mapbox-gl-native/build/android-arm-v7/Debug/../../../src/mbgl/util/tile_cover.cpp:112
Stack frame #04 pc 001342cc  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine std::__ndk1::chrono::duration<long long, std::__ndk1::ratio<1ll, 1000000000ll> >::operator+=(std::__ndk1::chrono::duration<long long, std::__ndk1::ratio<1ll, 1000000000ll> > const&) at /Users/ivo/git/mapbox-gl-native/mason_packages/osx-x86_64/android-ndk/arm-9-r12b/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/chrono:494
Stack frame #05 pc 00134158  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine stopwatch at /Users/ivo/git/mapbox-gl-native/build/android-arm-v7/Debug/../../../src/mbgl/util/stopwatch.cpp:14
Stack frame #06 pc 0012ea1c  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine void std::__ndk1::allocator<std::__ndk1::__hash_node<std::__ndk1::__hash_value_type<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, mapbox::geometry::value>, void*> >::construct<mapbox::geometry::value>(mapbox::geometry::value*) at /Users/ivo/git/mapbox-gl-native/mason_packages/osx-x86_64/android-ndk/arm-9-r12b/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/memory:1645
Stack frame #07 pc 00133d1c  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine mbgl::matrix::multiply(std::__ndk1::array<double, 16u>&, std::__ndk1::array<double, 16u> const&, std::__ndk1::array<double, 16u> const&) at /Users/ivo/git/mapbox-gl-native/build/android-arm-v7/Debug/../../../src/mbgl/util/mat4.cpp:315
Stack frame #08 pc 00133b2c  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine mbgl::matrix::rotate_z(std::__ndk1::array<double, 16u>&, std::__ndk1::array<double, 16u> const&, double) at /Users/ivo/git/mapbox-gl-native/build/android-arm-v7/Debug/../../../src/mbgl/util/mat4.cpp:273
Stack frame #09 pc 00137074  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine UnwrappedTileID at /Users/ivo/git/mapbox-gl-native/build/android-arm-v7/Debug/../../../src/mbgl/tile/tile_id.hpp:205 (discriminator 3)
Stack frame #10 pc 00136f58  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine bool std::__ndk1::__insertion_sort_incomplete<mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::$_1&, mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::ID*>(mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::ID*, mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::ID*, mbgl::util::(anonymous namespace)::tileCover(mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, mapbox::geometry::point<double> const&, int)::$_1&) at /Users/ivo/git/mapbox-gl-native/mason_packages/osx-x86_64/android-ndk/arm-9-r12b/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/type_traits:3308 (discriminator 3)
Stack frame #11 pc 0012f0b0  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine std::__ndk1::allocator<std::__ndk1::reference_wrapper<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const> >::deallocate(std::__ndk1::reference_wrapper<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const>*, unsigned int) at /Users/ivo/git/mapbox-gl-native/mason_packages/osx-x86_64/android-ndk/arm-9-r12b/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/memory:1636
Stack frame #12 pc 0012eff8  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine std::__ndk1::__tree<std::__ndk1::__value_type<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, unsigned int>, std::__ndk1::__map_value_compare<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::__value_type<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, unsigned int>, std::__ndk1::less<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, true>, std::__ndk1::allocator<std::__ndk1::__value_type<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, unsigned int> > >::__node_insert_unique(std::__ndk1::__tree_node<std::__ndk1::__value_type<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, unsigned int>, void*>*) at /Users/ivo/git/mapbox-gl-native/mason_packages/osx-x86_64/android-ndk/arm-9-r12b/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/__tree:1933
Stack frame #13 pc 00016baf  /system/lib/libc.so (__pthread_start(void*)+30)
Stack frame #14 pc 00014af3  /system/lib/libc.so (__start_thread+6)
```

**Comment by jfirebaugh** [comment]
@ivovandongen Changing visibility should work now, thanks for the catch.

That stack trace looks nonsensical. Could be a symbol mismatch. I really need to see the full logs  (#6220) and possibly help interpreting them.

**Comment by jfirebaugh** [comment]
The way I'd like to fix the race condition is to change the worker API to:

```
void setGeometryTileData(worker, data); // invoked in "fire and forget" fashion, not cancellable
Request startLayout(worker, ..., callback); // replaces parseGeometryTile, cancellable
Request continueLayout(worker, ..., callback); // replaces parsePendingGeometryTileLayers, cancellable
```

However, this depends on serial processing order of worker messages, so that we can guarantee that `setGeometryTileData` has completed by the time `startLayout` begins. We don't have right now: it's possible for `setGeometryTileData` to get dispatched to one thread and `startLayout` to get dispatched to another and start executing before `setGeometryTileData`.

So I'm going to work on fixing #1471, which if implemented in the way I have in mind, would solve this problem as well.

**Comment by jfirebaugh** [comment]
Changing the worker API is turning out to be difficult. In order to unblock other issues, I'm going to merge the changes so far and ticket the race condition and GeoJSON mutation for followups.

**Comment by jfirebaugh** [comment]
iOS build is failing because in the latest rebase I wiped out d77a13eb7320722c48c8a18240adf99615c4b85f with the deletion of `MGLBaseStyleLayer_Private.h`. @1ec5 @frederoni, where should `#import "NSPredicate+MGLAdditions.h"` be moved?

**Comment by incanus** [comment]
🎉

**Pull request merged**